packages/blade/src/components/Input/BaseInput/BaseInput.tsx

```diff
@@ -1,73 +1,167 @@
+  neccessityIndicator?: FormLabelProps['neccessityIndicator'];
 };
 
-export type BaseInputProps = LabelProps & {
+// TODO: need to abstract for generic use
+type InputValidationProps = {
   /**
-   * Placeholder text to be displayed inside the input field
+   * Help text for the input
    */
-  placeholder?: string;
+  helpText?: string;
   /**
-   * Type of Input Field to be rendered.
+   * Error text for the input
    *
-   * @default text
-   */
-  type?: 'text' | 'telephone' | 'email' | 'url' | 'numeric' | 'search';
-  /**
-   * Used to set the default value of input field when it's uncontrolled
+   * Renders when `validationState` is set to 'error'
    */
-  defaultValue?: string;
+  errorText?: string;
   /**
-   * The name of the input field.
+   * success text for the input
    *
-   * Useful in form submissions
+   * Renders when `validationState` is set to 'success'
    */
-  name?: string;
+  successText?: string;
   /**
-   * The callback function to be invoked when the value of the input field changes
+   * If `error`, the input is marked as invalid,
+   * and `invalid` attribute will be added
+   *
+   * If `success`, the input is marked as valid,
+   *
    */
-  onChange?: OnChange;
+  validationState?: 'success' | 'error' | 'none';
+};
+
+export type BaseInputProps = InputLabelProps &
+  InputValidationProps & {
+    /**
+     * ID that will be used for accessibility
+     */
+    id: string;
+    /**
+     * Placeholder text to be displayed inside the input field
+     */
+    placeholder?: string;
+    /**
+     * Type of Input Field to be rendered.
+     *
+     * @default text
+     */
+    type?: 'text' | 'telephone' | 'email' | 'url' | 'numeric' | 'search';
+    /**
+     * Used to set the default value of input field when it's uncontrolled
+     */
+    defaultValue?: string;
+    /**
+     * The name of the input field.
+     *
+     * Useful in form submissions
+     */
+    name?: string;
+    /**
+     * The callback function to be invoked when the value of the input field changes
+     */
+    onChange?: OnChange;
+    /**
+     * Used to turn the input field to controlled so user can control the value
+     */
+    value?: string;
+    /**
+     * Used to disable the input field
+     */
+    isDisabled?: boolean;
+    /**
+     * If true, the input is marked as required, and `required` attribute will be added
+     */
+    isRequired?: boolean;
+  };
+
+const useInput = ({
+  value,
+  defaultValue,
+  onChange,
+}: Pick<BaseInputProps, 'value' | 'defaultValue' | 'onChange'>): {
+  handleOnChange: HandleOnChange;
+} => {
+  if (value && defaultValue) {
+    throw new Error(
+      `[Blade Input]: Either 'value' or 'defaultValue' shall be passed. This decides if the input field is controlled or uncontrolled`,
+    );
+  }
+
+  const handleOnChange: HandleOnChange = React.useCallback(
+    ({ name, value }) => {
+      let _value = '';
+
+      if (getPlatformType() === 'react-native' && typeof value === 'string') {
```

Comment:
Can we move platform specific logic to .web / .native files?

Reply:
so at every step you kind of ask this question that shall we move this to a new file or same file with a basic check. In this case it's just 2-3 things which are diff rest the entire thing is same so putting a check is less noisy and less duplication